// type-utils/js/src/validation.js
var getTypeNames, reportFailure;

reportFailure = require("report-failure");

getTypeNames = require("./helpers").getTypeNames;

module.exports = function(TU) {
  var addValidatorType, configs, validateType, validateTypes, validateWithArray, validateWithFunction;
  configs = [];
  validateType = function(value, validator, keyPath) {
    var config, error, i, len;
    for (i = 0, len = configs.length; i < len; i++) {
      config = configs[i];
      if (config.isType(validator)) {
        return config.validate(value, validator, keyPath);
      }
    }
    error = TypeError("'validator' has an unexpected type.");
    return reportFailure(error, {
      key: keyPath,
      value: value,
      validator: validator
    });
  };
  validateTypes = function(obj, validators, keyPath) {
    var key, validator, value;
    if (validators == null) {
      return;
    }
    TU.assertType(validators, Object, keyPath);
    for (key in validators) {
      validator = validators[key];
      value = obj[key];
      if (keyPath != null) {
        key = keyPath + "." + key;
      }
      if (TU.isType(validator, Object)) {
        if (value == null) {
          continue;
        }
        TU.assertKind(value, Object, key);
        validateTypes(value, validator, key);
      } else {
        validateType(value, validator, key);
      }
    }
  };
  addValidatorType = function(config) {
    if (!(config.isType instanceof Function)) {
      throw TypeError("'config.isType' must be a Function");
    }
    if (!(config.validate instanceof Function)) {
      throw TypeError("'config.validate' must be a Function");
    }
    return configs.unshift(config);
  };
  validateWithArray = function(value, types, keyPath) {
    var error, typeNames;
    if (types.length === 0) {
      return;
    }
    if (types.length === 1) {
      return TU.assertType(value, types[0], keyPath);
    }
    if (TU.isType(value, types)) {
      return;
    }
    keyPath = keyPath != null ? "'" + keyPath + "'" : "This property";
    typeNames = getTypeNames(types);
    error = TypeError(keyPath + " must be a " + typeNames);
    return reportFailure(error, {
      key: keyPath,
      value: value,
      types: types
    });
  };
  addValidatorType({
    isType: function(value) {
      return TU.isType(value, Array);
    },
    validate: validateWithArray
  });
  validateWithFunction = function(value, validator, keyPath) {
    return validator(value, keyPath);
  };
  addValidatorType({
    isType: function(value) {
      return TU.isKind(value, Function);
    },
    validate: validateWithFunction
  });
  return {
    validateTypes: validateTypes,
    addValidatorType: addValidatorType
  };
};
